src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc

```diff
@@ -505,58 +525,72 @@ XdsResolver::XdsConfigSelector::XdsConfigSelector(
         route_action->max_stream_duration =
             resolver_->current_listener_.http_max_stream_duration;
       }
-      Match(
+      absl::StatusOr<std::vector<std::string>> status_or_cluster_names = Match(
```

Comment:
Using a `std::vector<>` here causes unnecessary allocation and makes the code less readable.  I think we're better off sticking with the previous approach of adding each cluster as we identify it, inside the lambdas.

I suggest adding a lambda like this on line 510 above:

```
auto maybe_add_cluster = [&](absl::string_view cluster_name) {
  if (clusters.find(cluster_name) != clusters.end()) return;
  auto cluster_state = resolver_->GetOrCreateClusterState(cluster_name);
  absl::string_view name = cluster_state->cluster_name();
  clusters.emplace(name, std::move(cluster_state));
};
```

Then we don't need any changes to this `Match()` call except for changing `MaybeAddCluster` to `maybe_add_cluster`.